### PR TITLE
Write contracts + pre/post conditions for all unsafe methods in duration

### LINKED ISSUE #72: Challenge 9: Safe abstractions for `core::time::Duration`
[Challenge 9: Safe abstractions for core::time::Duration](https://model-checking.github.io/verify-rust-std/challenges/0009-duration.html#challenge-9-safe-abstractions-for-coretimeduration)

### PR DESCRIPTION
**Pull request by sgpthomas**
Resolves #72 

We added invariants for Nanoseconds and Duration to match the safety conditions for those types. 

We add safety requirements to the following methods:
- `new`, `from_secs`, `from_millis`, `from_micros`, `from_nanos`, `as_secs`, `as_millis`, `as_micros`, `as_nanos`, `subsec_millis`, `subsec_micros`, `subsec_nanos`, `checked_add`, `checked_sub`, `checked_mul`, `checked_div`

We additionally add correctness conditions to the following methods:
- `from_secs`, `as_secs`, `subsec_millis`, `subsec_micros`, `subsec_nanos`, `as_millis`, `as_micros`

Support for `kani::Invariant` depends on #87. For the interim we implemented a proxy trait `TempInvariant` that exposes the same `is_safe` method. We will update this once #87 is merged.

While the safety check for `Duration::as_nanos()` succeeds, we ran into timeouts for `Duration::as_nanos()` when we tried to use a correctness contract and we're looking for advice on how to speed up that verification time. We were able to prove it for `u16::MAX`, but hit timeouts for larger numbers. 

We are unsure if the pre-conditions for `Duration::new()` are acceptable because they limit the range of values that you can call `Duration::new()` with. However, we think it's reasonable since we limit the values to values that don't panic. Let us know if this is a thing that we should change.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/time.rs [hunk 1]
```diff
@@ -37,12 +41,25 @@ const HOURS_PER_DAY: u64 = 24;
...
+
 #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
 #[repr(transparent)]
 #[rustc_layout_scalar_valid_range_start(0)]
 #[rustc_layout_scalar_valid_range_end(999_999_999)]
 struct Nanoseconds(u32);
 
+// We couldn't get the `kani::Invariant` trait to work, so here is a temporary
+// implementation that matches what that trait would be. Once pr #87 is merged,
+// we will update this to use `kani::Invariant`
```
Comment (celinval) [inline library/core/src/time.rs:56]: The trait is being added to `crate::ub_checks`.

File: library/core/src/time.rs [hunk 2]
```diff
@@ -1701,3 +1765,121 @@ impl Duration {
...
+    impl kani::Arbitrary for Duration {
+        fn any() -> Duration {
+            let secs = kani::any::<u64>();
+            let nanos = kani::any::<u32>();
+            kani::assume(nanos < NANOS_PER_SEC || secs.checked_add((nanos / NANOS_PER_SEC) as u64).is_some());
+            Duration::new(secs, nanos)
+        }
+    }
+
+    #[kani::proof_for_contract(Duration::new)]
```
Comment (celinval) [inline library/core/src/time.rs:1784]: do you mind adding a second harness for `duration_new` that does not use contract, and ensures that there's a panic instead? you can use `#[kani::should_panic]`. Thanks!

File: library/core/src/time.rs [hunk 3]
```diff
@@ -486,6 +522,8 @@ impl Duration {
     #[rustc_const_stable(feature = "duration_consts", since = "1.32.0")]
     #[must_use]
     #[inline]
+    #[requires(self.is_safe())]
```
Comment (celinval) [inline library/core/src/time.rs:525]: For safe functions, you should be able to assume that is the case. So this becomes redundant.

File: library/core/src/time.rs [hunk 4]
```diff
@@ -817,14 +876,19 @@ impl Duration {
...
+    #[ensures(|duration| rhs == 0 || duration.unwrap().is_safe())]
     #[rustc_const_stable(feature = "duration_consts_2", since = "1.58.0")]
     pub const fn checked_div(self, rhs: u32) -> Option<Duration> {
         if rhs != 0 {
             let (secs, extra_secs) = (self.secs / (rhs as u64), self.secs % (rhs as u64));
             let (mut nanos, extra_nanos) = (self.nanos.0 / rhs, self.nanos.0 % rhs);
             nanos +=
                 ((extra_secs * (NANOS_PER_SEC as u64) + extra_nanos as u64) / (rhs as u64)) as u32;
-            debug_assert!(nanos < NANOS_PER_SEC);
+            // TODO: why does this debug assert cause verification to timeout
```
Comment (celinval) [inline library/core/src/time.rs:888]: It's because division is hard for SAT solvers. How bad is the performance with it? We don't want to change the existing code.

**Review by celinval — COMMENTED** [review]
That's awesome! Thank you

File: library/core/src/time.rs (hunk 4 above)
Comment (sgpthomas) [reply library/core/src/time.rs:888]: I have left it running for over an hour without it finishing. I'll try for longer. What's the max time you would accept? But more fundamentally, I'm confused why this verification would succeed without the debug assert, but fails with the debug_assert when the debug assert is checking essentially the same condition. Do you have any idea why that might be?

Comment (celinval) [reply library/core/src/time.rs:864]: Good question. @tautschnig might have better insight.

This is not surprising to me though. This reminds me of https://github.com/model-checking/kani/issues/1351

File: library/core/src/time.rs [hunk 5]
```diff
@@ -95,8 +110,10 @@ impl Default for Nanoseconds {
 #[stable(feature = "duration", since = "1.3.0")]
 #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
 #[cfg_attr(not(test), rustc_diagnostic_item = "Duration")]
+#[cfg_attr(kani, derive(Invariant))]
 pub struct Duration {
     secs: u64,
+    // #[safety_constraint(*nanos.0 < NANOS_PER_SEC)]
```
[the remark this answers is not in the document]
Comment (cvick32) [reply library/core/src/time.rs]: It was commented out as we were experimenting with Invariant. Removed as it's no longer necessary.

File: library/core/src/time.rs [hunk 6]
```diff
@@ -705,6 +737,7 @@ impl Duration {
                   without modifying the original"]
     #[inline]
     #[rustc_const_stable(feature = "duration_consts_2", since = "1.58.0")]
+    #[ensures(|duration| !duration.is_some() || duration.unwrap().is_safe())]
```
Comment (celinval) [inline library/core/src/time.rs:740]:
```suggestion
    #[ensures(|duration| duration.is_none() || duration.unwrap().is_safe())]
```

File: library/core/src/time.rs [hunk 7]
```diff
@@ -761,6 +794,7 @@ impl Duration {
...
```
Comment (celinval) [inline library/core/src/time.rs:797]:
```suggestion
    #[ensures(|duration| duration.is_none() || duration.unwrap().is_safe())]
```

**Review by celinval — APPROVED** [review]
Awesome! Thanks

File: library/core/src/time.rs (hunk 2 above)
Comment (cvick32) [reply library/core/src/time.rs:1765]: Just added the should_panic harness.

File: library/core/src/time.rs (hunk 4 above)
Comment (tautschnig) [reply library/core/src/time.rs:861]: Yes, just one extra multiplication (division is turned into a multiplication) can be sufficient to make the problem move from works-in-a-few-minutes to runs-for-several-hours. You might want to change the post-condition so that it is disabled under cfg(kani).

File: library/core/src/time.rs [hunk 8]
```diff
@@ -37,12 +43,19 @@ const HOURS_PER_DAY: u64 = 24;
 #[unstable(feature = "duration_units", issue = "120301")]
 const DAYS_PER_WEEK: u64 = 7;
 
+
```
Comment (tautschnig) [inline library/core/src/time.rs:46]: Unnecessary change?

File: library/core/src/time.rs [hunk 9]
```diff
@@ -208,6 +222,9 @@ impl Duration {
     #[inline]
     #[must_use]
     #[rustc_const_stable(feature = "duration_consts_2", since = "1.58.0")]
+    // by definition of NANOS_PER_SEC, the checks for div by 0 cases are unreachable, but also unneeded
```
Comment (tautschnig) [inline library/core/src/time.rs:225]: I admit that I mostly find this comment confusing. It's a comment about a detail at a point where, as a reader, I am trying to get the bigger picture first. Does this comment really add value? Perhaps you should, instead, just spell this out as a precondition?

File: library/core/src/time.rs [hunk 10]
```diff
@@ -258,6 +277,8 @@ impl Duration {
     #[must_use]
     #[inline]
     #[rustc_const_stable(feature = "duration_consts", since = "1.32.0")]
+    // by definition of MILLIS_PER_SEC, the checks for div by 0 cases are unreachable, but also unneeded
```
Comment (tautschnig) [inline library/core/src/time.rs:280]: As above: should you perhaps just spell this out as a precondition, even though it is trivially satisfied (at the moment!)?

File: library/core/src/time.rs [hunk 11]
```diff
@@ -284,6 +305,8 @@ impl Duration {
...
```
Comment (tautschnig) [inline library/core/src/time.rs:308]: As above.

File: library/core/src/time.rs [hunk 12]
```diff
@@ -315,6 +338,8 @@ impl Duration {
...
```
Comment (tautschnig) [inline library/core/src/time.rs:341]: As above.

File: library/core/src/time.rs [hunk 13]
```diff
@@ -508,6 +534,7 @@ impl Duration {
     #[rustc_const_stable(feature = "duration_consts", since = "1.32.0")]
     #[must_use]
     #[inline]
+    #[ensures(|ms| *ms == self.nanos.0 / NANOS_PER_MILLI)]
```
Comment (tautschnig) [inline library/core/src/time.rs:537]: With the above comments (or perhaps preconditions) you'd want to do the same for `NANOS_PER_MILLI`.

File: library/core/src/time.rs [hunk 14]
```diff
@@ -531,6 +558,7 @@ impl Duration {
     #[rustc_const_stable(feature = "duration_consts", since = "1.32.0")]
     #[must_use]
     #[inline]
+    #[ensures(|ms| *ms == self.nanos.0 / NANOS_PER_MICRO)]
```
Comment (tautschnig) [inline library/core/src/time.rs:561]: With the above comments (or perhaps preconditions) you'd want to do the same for `NANOS_PER_MICRO`.

File: library/core/src/time.rs [hunk 15]
```diff
@@ -647,6 +678,7 @@ impl Duration {
                   without modifying the original"]
     #[inline]
     #[rustc_const_stable(feature = "duration_consts_2", since = "1.58.0")]
+    #[ensures(|duration| !duration.is_some() || duration.unwrap().is_safe())]
```
Comment (tautschnig) [inline library/core/src/time.rs:681]: For consistency with the below use `duration.is_none()`

File: library/core/src/time.rs [hunk 16]
```diff
@@ -208,6 +222,9 @@ impl Duration {
     #[inline]
     #[must_use]
     #[rustc_const_stable(feature = "duration_consts_2", since = "1.58.0")]
+    // by definition of NANOS_PER_SEC, the checks for div by 0 cases are unreachable, but also unneeded
+    #[requires(nanos < NANOS_PER_SEC || secs.checked_add((nanos / NANOS_PER_SEC) as u64).is_some())]
```
Comment (tautschnig) [inline library/core/src/time.rs:226]: This is a `safe` function, why do we have a `requires` clause?

**Review by tautschnig — CHANGES_REQUESTED** [review]
I don't think we should have anything other than `true` as a `requires` clause on safe functions.

File: library/core/src/time.rs (hunk 4 above)
Comment (cvick32) [reply library/core/src/time.rs:862]: I updated the post-condition to just `true` when running Kani, but it still times out for me locally. Not sure what else to try here.

File: library/core/src/time.rs (hunk 16 above)
Comment (tautschnig) [reply library/core/src/time.rs:225]: The function is not marked `unsafe`, which implies that it must not exhibit undefined behaviour for any input (per Rust's compile-time guarantees). A `panic` isn't undefined behaviour, so there isn't a need to guarantee that this does not happen. We may, however, still need to tweak something with regard to the `should_fail` predicate of Kani to make sure Kani isn't unhappy.

Comment (cvick32) [reply library/core/src/time.rs:225]: We can remove the `kani::assume` from `Duration::any` because we don't need the assumption if we've removed the precondition. Then we can add a `should_panic` to the contract proof for `Duration::new()`.  Like so: 

```rust
impl kani::Arbitrary for Duration {
        fn any() -> Duration {
            let secs = kani::any::<u64>();
            let nanos = kani::any::<u32>();
            //kani::assume(nanos < NANOS_PER_SEC || secs.checked_add((nanos / NANOS_PER_SEC) as u64).is_some());
            Duration::new(secs, nanos)
        }
    }

    #[kani::proof_for_contract(Duration::new)]
    #[kani::should_panic]
    fn duration_new() {
        let secs = kani::any::<u64>();
        let nanos = kani::any::<u32>();
        let _ = Duration::new(secs, nanos);
    }
```

Kani is happy with this contract, but now the following contracts fail because a call to `Duration::any()` can now panic: 
```
Verification failed for - time::duration_verify::duration_checked_mul
Verification failed for - time::duration_verify::duration_checked_sub
Verification failed for - time::duration_verify::duration_checked_add
Verification failed for - time::duration_verify::duration_as_nanos
Verification failed for - time::duration_verify::duration_as_micros
Verification failed for - time::duration_verify::duration_as_millis
Verification failed for - time::duration_verify::duration_subsec_nanos
Verification failed for - time::duration_verify::duration_subsec_micros
Verification failed for - time::duration_verify::duration_subsec_millis
Verification failed for - time::duration_verify::duration_as_secs
```

We can add `should_panic` to all of these contract proofs, but I'm not sure if that's the best option.

File: library/core/src/time.rs (hunk 4 above)
Comment (cvick32) [reply library/core/src/time.rs]: I added the `cfg_attr` to the `debug_assert` inside of `checked_div` and Kani was able to complete the proof. Is this addition to the code itself allowed as it doesn't change the behavior when running without Kani?

File: library/core/src/time.rs [hunk 17]
```diff
@@ -1698,3 +1728,131 @@ impl Duration {
...
+        let nanos = kani::any::<u64>();
+        let _ = Duration::from_nanos(nanos);
+    }
+
+    #[kani::proof_for_contract(Duration::as_secs)]
+    #[kani::should_panic]
+    fn duration_as_secs() {
+        let dur = kani::any::<Duration>();
+        let _ = dur.as_secs();
+    }
```
Comment (tautschnig) [inline library/core/src/time.rs:1783]: In keeping with what we recommend at https://model-checking.github.io/kani/reference/attributes.html#kanishould_panic I'd say it would be great to have a pair of harnesses for each contract: one that either firmly constrains the input to one that will cause a `panic`, or maybe just leave it unconstrained as is; and a second one constrains the input (via assumptions or the likes) to cases where no panic is expected.

**Review by tautschnig — APPROVED** [review]

File: library/core/src/time.rs (hunk 17 above)
Comment (cvick32) [reply library/core/src/time.rs]: Cool, just added the matching non-panicking harnesses for all harnesses that use `Duration::any()`. I also added a function `safe_duration` that will always return a `Duration` without panicking.

File: library/core/src/time.rs [hunk 18]
```diff
@@ -1698,3 +1728,213 @@ impl Duration {
...
+            let secs = kani::any::<u64>();
+            let nanos = kani::any::<u32>();
+            Duration::new(secs, nanos)
+        }
+    }
+
+    fn safe_duration() -> Duration {
+        let secs = kani::any::<u64>();
+        let nanos = kani::any::<u32>();
+        kani::assume(nanos < NANOS_PER_SEC || secs.checked_add((nanos / NANOS_PER_SEC) as u64).is_some());
```
Comment (feliperodri) [inline library/core/src/time.rs:1749]: Why do we need this assumption in the harness?

Comment (cvick32) [reply library/core/src/time.rs:1749]: This assumption constrains the inputs to `Duration::new()`, so that the call never `panics`. The constraints were taken from the implementation of `Duration::new()`:

```rust
pub const fn new(secs: u64, nanos: u32) -> Duration {
        if nanos < NANOS_PER_SEC {
            // SAFETY: nanos < NANOS_PER_SEC, therefore nanos is within the valid range
            Duration { secs, nanos: unsafe { Nanoseconds(nanos) } }
        } else {
            let secs = secs
                .checked_add((nanos / NANOS_PER_SEC) as u64)
                .expect("overflow in Duration::new");
            let nanos = nanos % NANOS_PER_SEC;
            // SAFETY: nanos % NANOS_PER_SEC < NANOS_PER_SEC, therefore nanos is within the valid range
            Duration { secs, nanos: unsafe { Nanoseconds(nanos) } }
        }
    }
```

You can see that the only time where `Duration::new()` panics is when `secs.checked_add((nanos / NANOS_PER_SEC) as u64)` is `None`. 

We added this non-panicking harness so that we could have panicking and non-panicking contracts.

Comment (feliperodri) [reply library/core/src/time.rs:1749]: Why shouldn't we capture this condition as part of the requires clause for `Duration::new`?

Comment (cvick32) [reply library/core/src/time.rs:1749]: Micheal left comments above stating that `safe` functions shouldn't have `requires` clauses. So, we moved the assumption to the guarded function `safe_duration` which guarantees that `Duration::new()` won't panic.
